llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -612,6 +612,118 @@ Instruction *emitSpecConstantRecursive(Type *Ty, Instruction *InsertBefore,
 
 } // namespace
 
+/// Function creates load instruction from the given Buffer by the given Offset.
+/// Function returns the value of load instruction.
+static Value *createLoadFromBuffer(CallInst *CI, Value *Buffer, size_t Offset,
+                                   Type *SCType) {
+  LLVMContext &C = CI->getContext();
+  Type *Int8Ty = Type::getInt8Ty(C);
+  Type *Int32Ty = Type::getInt32Ty(C);
+  GetElementPtrInst *GEP = GetElementPtrInst::Create(
+      Int8Ty, Buffer, {ConstantInt::get(Int32Ty, Offset, false)}, "gep", CI);
+
+  Instruction *BitCast = nullptr;
+  if (SCType->isIntegerTy(1)) // No bitcast to i1 before load
+    BitCast = GEP;
+  else
+    BitCast = new BitCastInst(
+        GEP, PointerType::get(SCType, GEP->getAddressSpace()), "bc", CI);
+
+  // When we encounter i1 spec constant, we still load the whole byte
+  Value *Load = new LoadInst(SCType->isIntegerTy(1) ? Int8Ty : SCType, BitCast,
+                             "load", CI);
+  if (SCType->isIntegerTy(1)) // trunc back to i1 if necessary
+    Load = CastInst::CreateIntegerCast(Load, SCType, /* IsSigned */ false,
+                                       "tobool", CI);
+
+  return Load;
+}
+
+/// Function tries to dig out the initializer from the given CallInst to
+/// SpecConst function. ArgIndex is the expected index of the function operand
+/// leading to the initializer.
+///
+/// Examples:
+/// 1)
+///   %"spec_id" = type { i32 }
+///   @value = internal addrspace(1) constant %"spec_id" { i32 123 }, align 4
+///   call spir_func i32 @sycl_getScalar2020SpecConst(%1, @value, %2)
+///
+/// 2)
+///   %"spec_id" = type { %A }
+///   %A = type { i32 }
+///   @value = constant %"spec_id" { %A { i32 1 } }, align 4
+///   call spir_func void @getCompositeSpecConst(%1, %2, @value, %3)
+static Constant *tryGetSpecConstInitializerFromCI(CallInst *CI,
+                                                  unsigned ArgIndex) {
+  assert(CI->getNumOperands() > ArgIndex && "ArgIndex is out of boundary");
```

Comment:
`CallInst::getArgOperand()` likely has similar assertion, I think we can omit it here

```suggestion
```